presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java

```diff
@@ -686,156 +698,109 @@ public static double round(@SqlType(StandardTypes.DOUBLE) double num, @SqlType(B
-            return num;
-        }
-        if (num < 0) {
-            return -roundShortDecimal(-num, rescaleFactor, inputScale);
+            long rescaleFactor = Decimals.longTenToNth(numScale.intValue());
+            long remainder = num % rescaleFactor;
+            long remainderBoundary = rescaleFactor >> 1;
+            int roundUp = remainder >= remainderBoundary ? 1 : 0;
+            return num / rescaleFactor + roundUp;
         }
 
-        long remainder = num % rescaleFactor;
-        long remainderBoundary = rescaleFactor >> 1;
-        int roundUp = remainder >= remainderBoundary ? 1 : 0;
-        return num / rescaleFactor + roundUp;
-    }
-
-    private static List<Object> decimalRoundLongExtraParameters(SpecializeContext context)
-    {
-        long scale = context.getLiteral("num_scale");
-        BigInteger rescaleFactor = bigIntegerTenToNth((int) scale);
-        return ImmutableList.of(rescaleFactor, scale);
-    }
-
-    public static Slice roundLongDecimal(Slice numSlice, BigInteger rescaleFactor, long inputScale)
-    {
-        BigInteger num = decodeUnscaledValue(numSlice);
-        if (num.signum() == 0) {
-            return encodeUnscaledValue(0);
-        }
-        if (inputScale == 0) {
-            return encodeUnscaledValue(num);
-        }
-        if (num.signum() < 0) {
-            return encodeUnscaledValue(roundLongDecimal(num.negate(), rescaleFactor).negate());
+        @LiteralParameters({"p", "s", "rp", "rs"})
+        @SqlType("decimal(rp, rs)")
+        @Constraint(variable = "rp", expression = "min(38, p - s + min(1, s))")
+        @Constraint(variable = "rs", expression = "0")
+        public static Slice roundLong(@LiteralParameter("s") Long numScale, @SqlType("decimal(p, s)") Slice num)
+        {
+            if (numScale == 0) {
+                return num;
+            }
+            return rescale(num, -numScale.intValue());
         }
-        return encodeUnscaledValue(roundLongDecimal(num, rescaleFactor));
     }
 
-    private static BigInteger roundLongDecimal(BigInteger num, BigInteger rescaleFactor)
-    {
-        BigInteger[] divideAndRemainder = num.divideAndRemainder(rescaleFactor);
-        BigInteger roundUp = divideAndRemainder[1].compareTo(rescaleFactor.shiftRight(1)) >= 0 ? ONE : ZERO;
-        return divideAndRemainder[0].add(roundUp);
-    }
+    @ScalarFunction("round")
+    @Description("round to nearest integer")
+    public static final class RoundN
+    {
+        @LiteralParameters({"p", "s", "rp"})
+        @SqlType("decimal(rp, s)")
+        @Constraint(variable = "rp", expression = "min(38, p + 1)")
+        public static long roundNShort(
+                @LiteralParameter("p") Long numPrecision,
+                @LiteralParameter("s") Long numScale,
+                @SqlType("decimal(p, s)") long num,
+                @SqlType(BIGINT) long decimals)
+        {
+            if (num == 0 || numPrecision - numScale + decimals <= 0) {
```

Comment:
`numPrecision - numScale + decimals`: what case is this?
It only applies when `numPrecision - numScale = decimals = 0`. Please provide some explanation